### PR TITLE
java.sql.Time & java.sql.Date

### PR DESCRIPTION
**Pull request by zhangshen023**
**What this PR does**:
java class java.sql.Time & java.sql.Date refer to hessian.Time & hessian.Date
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
can use hessian.Time & hessian.Date in go for communicating with java.sql.Time & java.sql.Date in java
```release-note

```

### CODE REVIEW

**Comment by AlexStocks** [comment]
@zhangshen023 pls fix the travis failure.

**Comment by zhangshen023** [comment]
> @zhangshen023 pls fix the travis failure.

ok, i have a try

**Comment by zhangshen023** [comment]
ok,i will make a try

File: java_sql_time.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+func (d *Date) time() time.Time {
+	return d.Time
+}
+
+func (d *Date) setTime(time time.Time) {
+	d.Time = time
+}
+
```
Comment (AlexStocks) [inline java_sql_time.go:53]: add comment

File: java_sql_time.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+func (Date) JavaClassName() string {
+	return "java.sql.Date"
+}
+
+func (d *Date) ValueOf(dateStr string) error {
```
Comment (AlexStocks) [inline java_sql_time.go:58]: add comment. especially u should introduce u date format is "2006-01-02".

File: java_sql_time.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	time, err := time.Parse("2006-01-02", dateStr)
+	if err != nil {
+		return err
+	}
+	d.Time = time
+	return nil
+}
+
+func (d *Date) Year() int {
```
Comment (AlexStocks) [inline java_sql_time.go:67]: add comment or "// nolint"

File: java_sql_time.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	return d.Time.Year()
+}
+
+func (d *Date) Month() time.Month {
+	return d.Time.Month()
+}
+
+func (d *Date) Day() int {
```
Comment (AlexStocks) [inline java_sql_time.go:75]: add comment or "// nolint"

File: java_sql_time.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	d.Time = time
+	return nil
+}
+
+func (d *Date) Year() int {
+	return d.Time.Year()
+}
+
+func (d *Date) Month() time.Month {
```
Comment (AlexStocks) [inline java_sql_time.go:71]: add comment or "// nolint"

File: java_sql_time.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+func (d *Date) Day() int {
+	return d.Time.Day()
+}
+
+type Time struct {
+	time.Time
+}
+
+func (Time) JavaClassName() string {
```
Comment (AlexStocks) [inline java_sql_time.go:83]: add comment

File: java_sql_time.go [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+	return "java.sql.Time"
+}
+
+func (t Time) time() time.Time {
```
Comment (AlexStocks) [inline java_sql_time.go:87]: add comment or "// nolint"

File: java_sql_time.go [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+	return t.Time
+}
+
+func (t *Time) Hours() int {
```
Comment (AlexStocks) [inline java_sql_time.go:91]: add comment or "// nolint"

File: java_sql_time.go [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+	return t.Time.Hour()
+}
+
+func (t *Time) Minutes() int {
+	return t.Time.Minute()
+}
+
+func (t *Time) Seconds() int {
```
Comment (AlexStocks) [inline java_sql_time.go:99]: add comment or "// nolint"

File: java_sql_time.go [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+func (t Time) time() time.Time {
+	return t.Time
+}
+
+func (t *Time) Hours() int {
+	return t.Time.Hour()
+}
+
+func (t *Time) Minutes() int {
```
Comment (AlexStocks) [inline java_sql_time.go:95]: add comment or "// nolint"

File: java_sql_time.go [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+func (t *Time) Seconds() int {
+	return t.Time.Second()
+}
+
+func (t *Time) setTime(time time.Time) {
+	t.Time = time
+}
+
+func (t *Time) ValueOf(timeStr string) error {
```
Comment (AlexStocks) [inline java_sql_time.go:107]: add comment or "// nolint"

File: java_sql_time.go [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+	if err != nil {
+		return err
+	}
+	t.Time = time
+	return nil
+}
+
+var javaSqlTimeTypeMap = make(map[string]reflect.Type, 16)
+
+func SetJavaSqlTimeSerialize(time JavaSqlTime) {
```
Comment (AlexStocks) [inline java_sql_time.go:118]: add comment

File: java_sql_time.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+	SetSerializer(name, JavaSqlTimeSerializer{})
+	//RegisterPOJO(time)
+	javaSqlTimeTypeMap[name] = typ
+}
+
+func getJavaSqlTimeSerialize(name string) reflect.Type {
+	return javaSqlTimeTypeMap[name]
+}
+
+type JavaSqlTimeSerializer struct {
```
Comment (AlexStocks) [inline java_sql_time.go:130]: add comment or "// nolint"

File: java_sql_time.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+func (JavaSqlTimeSerializer) EncObject(e *Encoder, vv POJO) error {
```
Comment (AlexStocks) [inline java_sql_time.go:133]: add comment

File: java_sql_time.go [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+			return perrors.WithStack(err)
+		}
+
+		i = len(e.classInfoList)
+		e.classInfoList = append(e.classInfoList, clsDef)
+		e.buffer = append(e.buffer, clsDef.buffer...)
+		e.buffer = e.buffer[0 : len(e.buffer)-1]
+	}
+
+	if idx == -1 {
```
Comment (AlexStocks) [inline java_sql_time.go:187]: why using another if-clause as its condition is the same as the above if?

File: java_sql_time_test.go [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package hessian
+
+import (
+	"github.com/stretchr/testify/assert"
```
Comment (AlexStocks) [inline java_sql_time_test.go:21]: pls split the if clause.

File: java_sql_time.go (hunk 15 above)
Comment (zhangshen023) [reply java_sql_time.go:187]: optimized

File: java_sql_time_test.go (hunk 16 above)
Comment (zhangshen023) [reply java_sql_time_test.go:21]: Can you tell me in more detail

**Comment by wongoo** [comment]
NOT AGREE to add the Date&Time definition in hessian package,  and there already java8_time package in which it contains date and time definition.  If it's really necessary , suggestion to move to a new package `java_sql_time`.

**Comment by zhangshen023** [comment]
> java_sql_time

good suggestion.

**Comment by zhangshen023** [comment]
> NOT AGREE to add the Date&Time definition in hessian package, and there already java8_time package in which it contains date and time definition. If it's really necessary , suggestion to move to a new package `java_sql_time`.

good suggestion.

File: java_sql_time_test.go (hunk 16 above)
Comment (AlexStocks) [reply java_sql_time_test.go]: what  I mean is u should split the import into two parts as other dubbogo files.

Comment (zhangshen023) [reply java_sql_time_test.go]: > what I mean is u should split the import into two parts as other dubbogo files.

ok

File: java_sql_time.go [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+		return nil, err
+	}
+	date, err := d.decDate(int32(tag))
+	if err != nil {
+		return nil, perrors.WithStack(err)
+	}
+	sqlTime := vRef.Interface()
+
+	result, ok := sqlTime.(java_sql_time.JavaSqlTime)
+	result.SetTime(date)
```
Comment (wongoo) [inline java_sql_time.go:147]: set after checking ok

File: java_sql_time.go [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+import (
+	"github.com/apache/dubbo-go-hessian2/java_sql_time"
+)
+
+func init() {
+	RegisterPOJO(&java_sql_time.Date{})
+	RegisterPOJO(&java_sql_time.Time{})
+}
+
+var javaSqlTimeTypeMap = make(map[string]reflect.Type, 16)
```
Comment (wongoo) [inline java_sql_time.go:38]: javaSqlTimeTypeMap is unused, is it necessary?

File: java_sql_time.go [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go-hessian2/java_sql_time"
+)
+
+func init() {
+	RegisterPOJO(&java_sql_time.Date{})
+	RegisterPOJO(&java_sql_time.Time{})
```
Comment (wongoo) [inline java_sql_time.go:35]: should JavaSqlTimeSerializer be registered here?

File: java_sql_time.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+	javaSqlTimeTypeMap[name] = typ
+}
+
+// nolint
+func getJavaSqlTimeSerialize(name string) reflect.Type {
+	return javaSqlTimeTypeMap[name]
+}
+
+// JavaSqlTimeSerializer used to encode & decode java.sql.Time & java.sql.Date
+type JavaSqlTimeSerializer struct {
```
Comment (wongoo) [inline java_sql_time.go:54]: move all code to package `java_sql_time`, only keep the init func in hessian package

File: java_sql_time.go (hunk 18 above)
Comment (zhangshen023) [reply java_sql_time.go:38]: LGTM

File: java_sql_time.go (hunk 19 above)
Comment (zhangshen023) [reply java_sql_time.go:35]: LGTM

**Review by AlexStocks — APPROVED** [review]

File: java_sql_time/java_sql_time.go [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+package java_sql_time
+
+import "time"
+
+type JavaSqlTime interface {
+	// ValueOf parse time string which format likes '2006-01-02 15:04:05'
+	ValueOf(timeStr string) error
+	// SetTime for decode time
+	SetTime(time time.Time)
+	JavaClassName() string
```
Comment (zouyx) [inline java_sql_time/java_sql_time.go:27]: Should fix this problem?

[the remark this answers is not in the document]
Comment (zhangshen023) [reply java_sql_time/java_sql_time.go:27]: IMO,may cause circular reference. The hierarchy of package java_sql_time which struct 'JavaSqlTime' is kept is lower than interface Pojo's.
image

Comment (zhangshen023) [reply java_sql_time/java_sql_time.go:27]: pls see https://github.com/apache/dubbo-go-hessian2/pull/219#discussion_r479706008

**Review by zouyx — APPROVED** [review]

**Pull request merged**